### PR TITLE
Series of tests for AD

### PR DESCRIPTION
**Pull request by theogf**
I started to re-adapt the tests given the new structure and make them look a bit nicer.
Right now the idea is to compare the results (when possible) with FiniteDifferences.jl
The tests are quite generic and changeable.
~~The idea is to test differentiability given the inputs for all ADs and to test the differentiability given the kernel parameters only given Zygote. (other kernels would require more complex solutions)~~
The tests functions for AD are contained in `test/utils_AD.jl`.
A first function needs to be called with FiniteDifferences.jl to check that it runs without errors (eliminate one source of error when comparing)
A second function will compare a AD (so far implemented for Zygote, ForwardDiff and ReverseDiff) with a kernelfunction which may take arguments. The test setup for a kernel would look like this:
```
kfunction = (l -> transform(SEKernel(), first(l)))
args = [2.0]
v = test_FiniteDiff(kfunction, args)
if !v.anynonpass
    for AD in [:Zygote, :ForwardDiff, :ReverseDiff]
        test_AD(AD, kfunction, args)
    end
end
```
where `k = kernelfunction(args)` and args is a vector
Alternatively for no args :
```
kfunction = SqExponentialKernel
v = test_FiniteDiff(kfunction)
if !v.anynonpass
    for AD in [:Zygote, :ForwardDiff, :ReverseDiff]
        test_AD(AD, kfunction)
    end
end
```

Are tested :
- `kappa` for `SimpleKernel`
- `k(x, y)` given `k`, `x` AND `y`
- `kernelmatrix(A, B)` for both `RowVecs` and `ColVecs`, also given `A`, `B` and `k` (I use `sum` to get a scalar on it)
` `kernelmatrix(A)` same as above

Are added to make more tests pass:
- Indirection to `_map` from `Base.map` for avoid `Zygote` bug with `map` on `AbstractVector`
- More adjoints for `DotProduct` and `Sinus`
- Fixes for some function to be able to differentiate through (especially Matern Kernel)

### CODE REVIEW

**Comment by theogf** [comment]
One major issue is currently #113. And I have no idea how to solve it

**Comment by willtebbutt** [comment]
To be honest, I was planning to address this at the same time as #88.

In my opinion the correct way to do this is to incorporate differentiability tests in with the standard test suite, as opposed to having a collection of tests in a separate location.

No objection to this going in for now to get us in roughly the right place, but I don't think it's the correct solution long-term.

**Comment by theogf** [comment]
So you propose to integrate each all the AD test separately for each type of kernel in their test file?
It sounds good to me, but what I show for now is what do we even want to test. This is a very rough draft, also to explore what is currently failing.

**Comment by willtebbutt** [comment]
> So you propose to integrate each all the AD test separately for each type of kernel in their test file?

Yup.

> but what I show for now is what do we even want to test.

We want to test _everything_ in the public-facing API for every kernel that we write tests for.

> This is a very rough draft, also to explore what is currently failing.

👍

**Comment by theogf** [comment]
@willtebbutt Would you be satisfied by these tests? I would lay them out all over the tests but I prefer to check with you first :P

**Comment by willtebbutt** [comment]
It's definitely a reasonable starting point. We shouldn't really be testing gradients, rather we should be testing pushforwards / pullbacks. Unfortunately ForwardDiff and ReverseDiff don't expose these in their public APIs, so I'll need to do some digging after NeurIPS.

I would propose you merge this, because it's a definitely improvement over what we currently have, and we revisit after NeurIPS when refactoring the test suite.

**Comment by theogf** [comment]
Coverage has reduced because of some additional adjoints and functions that were needed for AD. Once I put all the tests in place it should actually go up :)

**Comment by theogf** [comment]
So I spread all the tests to the BaseKernels, Kernels and Transforms and when tests were failing I simply left out the AD in question and left a @test_broken with a comment. I listed all the failures in #116

**Comment by theogf** [comment]
I was also unable to find the proper adjoint for the pairwise function with the `Sinus` metric

**Comment by theogf** [comment]
Tests are failing but not because of new tests (some issue with the FBM kernel). Other than that it should be ready to merge

File: src/distances/delta.jl [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 
-@inline function Distances._evaluate(::Delta,a::AbstractVector{T},b::AbstractVector{T}) where {T}
+@inline function Distances._evaluate(::Delta, a::AbstractVector, b::AbstractVector) where {T}
     @boundscheck if length(a) != length(b)
         throw(DimensionMismatch("first array has length $(length(a)) which does not match the length of the second, $(length(b))."))
     end
     return a == b
 end
 
+Distances.result_type(::Delta, Ta::Type, Tb::Type) = promote_type(Ta, Tb)
```
Comment (devmotion) [inline src/distances/delta.jl:11]: This won't specialize on the types `Ta` and `Tb` (see https://docs.julialang.org/en/v1/manual/performance-tips/#Be-aware-of-when-Julia-avoids-specializing-1).

File: src/distances/dotproduct.jl [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 
-@inline function Distances._evaluate(::DotProduct, a::AbstractVector{T}, b::AbstractVector{T}) where {T}
+@inline function Distances._evaluate(::DotProduct, a::AbstractVector, b::AbstractVector)
     @boundscheck if length(a) != length(b)
         throw(DimensionMismatch("first array has length $(length(a)) which does not match the length of the second, $(length(b))."))
     end
     return dot(a,b)
 end
 
+Distances.result_type(::DotProduct, Ta::Type, Tb::Type) = promote_type(Ta, Tb)
```
Comment (devmotion) [inline src/distances/dotproduct.jl:11]: Same here.

File: src/distances/sinus.jl [hunk 3]
```diff
@@ -8,7 +8,9 @@ Distances.parameters(d::Sinus) = d.r
 @inline (dist::Sinus)(a::AbstractArray, b::AbstractArray) = Distances._evaluate(dist, a, b)
 @inline (dist::Sinus)(a::Number, b::Number) = abs2(sinpi(a - b) / first(dist.r))
 
-@inline function Distances._evaluate(d::Sinus, a::AbstractVector{T}, b::AbstractVector{T}) where {T}
+Distances.result_type(::Sinus{T}, Ta::Type, Tb::Type) where {T} = promote_type(T, Ta, Tb)
```
Comment (devmotion) [inline src/distances/sinus.jl:11]: And here.

File: src/utils.jl [hunk 4]
```diff
@@ -1,5 +1,7 @@
 hadamard(x, y) = x .* y
 
+loggamma(x) = first(logabsgamma(x))
```
Comment (devmotion) [inline src/utils.jl:3]: There's no reason to define `loggamma` other than avoiding the type piracy when defining an adjoint for `logabsgamma`, I guess?

File: src/zygote_adjoints.jl [hunk 5]
```diff
@@ -1,9 +1,88 @@
...
+
+@adjoint function pairwise(s::Sinus, X::AbstractMatrix, Y::AbstractMatrix; dims=2)
+    D = pairwise(d, X, Y; dims = dims)
+    throw(error("Sinus metric has no defined adjoint for now... PR welcome!"))
+end
+
+@adjoint function pairwise(s::Sinus, X::AbstractMatrix; dims=2)
+  D = pairwise(d, X; dims = dims)
+  throw(error("Sinus metric has no defined adjoint for now... PR welcome!"))
+end
```
Comment (devmotion) [inline src/zygote_adjoints.jl:70]: Maybe just not define an adjoint?

File: src/zygote_adjoints.jl [hunk 6]
```diff
@@ -1,9 +1,88 @@
...
+
+@adjoint function kappa(κ::MaternKernel, d::Real)
+    ν = first(κ.ν)
+    val, grad = pullback(_matern, ν, d)
+    return ((iszero(d) ? one(d) : val),
+    Δ -> begin
+        ∇ = grad(Δ)
+        return ((ν = [∇[1]],), iszero(d) ? zero(d) : ∇[2])
+    end)
+end
```
Comment (devmotion) [inline src/zygote_adjoints.jl:84]: Why is this needed? The definition of `kappa(::MaternKernel, ::Real)` seems quite simple, so I would have assumed that AD just works?

File: test/runtests.jl [hunk 7]
```diff
@@ -1,16 +1,15 @@
...
-using FiniteDifferences
-using Flux
 using Kronecker
 using LinearAlgebra
 using PDMats
 using Random
 using SpecialFunctions
 using Test
-using Zygote
+using Flux: params, Chain, Dense
```
Comment (devmotion) [inline test/runtests.jl:9]: Is it not possible to keep the old `using Flux` import?

File: test/utils_AD.jl [hunk 8]
```diff
@@ -1,39 +1,111 @@
-allapprox(x,y,tol=1e-8) = all(isapprox.(x,y,atol=tol))
-FDM = central_fdm(5,1)
 
+FDM = FiniteDifferences.central_fdm(5, 1)
```
Comment (devmotion) [inline test/utils_AD.jl:2]: It seems `FDM` is only used once in the definition of `gradient(::Val{:FiniteDiff}`, ...)`, so maybe just move it there?

File: test/utils_AD.jl [hunk 9]
```diff
@@ -1,39 +1,111 @@
...
 
-function kappa_fdm(k::Kernel,d::Real)
-    first(FiniteDifferences.grad(FDM,x->kappa(k,x),d))
+function gradient(::Val{:ReverseDiff}, f::Function, args)
+    ReverseDiff.gradient(f, args)
 end
 
+function gradient(::Val{:FiniteDiff}, f::Function, args)
+    first(FiniteDifferences.grad(FDM, f, args))
+end
```
Comment (devmotion) [inline test/utils_AD.jl:27]: Maybe drop the `::Function` annotation and make it the first argument. That would allow to specify `f` with the `do` syntax and be consistent with the general recommendations for positions of input arguments.

**Comment by devmotion** [comment]
> Tests are failing but not because of new tests (some issue with the FBM kernel).

I'm surprised about these test failures - it has to be related to the PR in some way, otherwise they wouldn't have passed all the time on master? Maybe try to set a seed (if we haven't fixed one yet)?

**Comment by willtebbutt** [comment]
> I'm surprised about these test failures - it has to be related to the PR in some way, otherwise they wouldn't have passed all the time on master? Maybe try to set a seed (if we haven't fixed one yet)?

Yeah, would definitely be nice to resolve this before merging.

File: src/distances/delta.jl (hunk 1 above)
Comment (theogf) [reply src/distances/delta.jl:11]: Distances do this for determining the type of the allocated matrix in `pairwise!`, if this is not defined it defaults to `Float64` and therefore breaks AD for ForwardDiff and others :
https://github.com/JuliaStats/Distances.jl/blob/f69f7888c92458ae671c893d079ecf5fc8d8accd/src/generic.jl#L35
https://github.com/JuliaStats/Distances.jl/blob/f69f7888c92458ae671c893d079ecf5fc8d8accd/src/generic.jl#L203

File: src/utils.jl (hunk 4 above)
Comment (theogf) [reply src/utils.jl:3]: Is that really type piracy? `logabsgamma` return a value and a sign, but we are only interested in the value so a wrapper solves this problems and allows to create a sensible adjoint.

File: src/zygote_adjoints.jl (hunk 5 above)
Comment (theogf) [reply src/zygote_adjoints.jl:70]: Sure! I just could not figure out the algebraic formulation of the adjoint for the Sinus

File: src/zygote_adjoints.jl (hunk 6 above)
Comment (theogf) [reply src/zygote_adjoints.jl:84]: It is needed for the edge case where `x = 0`, `matern` is defined as a `iszero(d) ? one(d) : _matern(d)` and `Zygote` will just return nothing on this branch kind of thing. It breaks the backpropagation.

File: test/runtests.jl (hunk 7 above)
Comment (theogf) [reply test/runtests.jl:9]: Sure I can export Flux generally

File: test/utils_AD.jl (hunk 8 above)
Comment (theogf) [reply test/utils_AD.jl:2]: Yes but it is reused everytime the gradient is called. I wanted to avoid useless allocations, I can make it a const!

File: test/utils_AD.jl (hunk 9 above)
Comment (theogf) [reply test/utils_AD.jl:27]: Oh nice I did not think about it this way. I will change it.

**Comment by theogf** [comment]
I will try to figure out the issue with FBM, but I thought this was maybe irrelevant to this PR ?

**Comment by theogf** [comment]
Can I merge this?

**Comment by theogf** [comment]
Ah maybe before I can put all dependencies for the test directly in a `Project.toml`

File: src/distances/delta.jl (hunk 1 above)
Comment (devmotion) [reply src/distances/delta.jl:11]: Sure, we don't want to restrict it to `Float64` only. My comment was just that Julia actually won't compile specialized versions of this function for different input types since you used `Ta::Type` instead of `::Type{Ta}`.

File: src/utils.jl (hunk 4 above)
Comment (devmotion) [reply src/utils.jl:3]: No, `loggamma` is no type piracy. I was just wondering why you define the function at all and not just define a custom adjoint for `logabsgamma` that outputs `nothing` for the second argument - and my assumption was that you didn't define one because you wanted to avoid this type piracy (of defining an adjoint for a function that we don't own).

File: src/zygote_adjoints.jl (hunk 6 above)
Comment (devmotion) [reply src/zygote_adjoints.jl:74]: Does `ifelse(iszero(d), one(d), _matern(d))` fix the Zygote issue? I've seen this pattern quite a bit in the Zygote source code, so maybe it's helpful in such cases?

File: test/Project.toml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+[deps]
+Distances = "b4f34e82-e78d-54a5-968a-f98e89d6e8f7"
+FiniteDifferences = "26cc04aa-876d-5657-8c51-4c34ba976000"
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
+KernelFunctions = "ec8451be-7e33-11e9-00cf-bbf324bd1392"
```
Comment (devmotion) [inline test/Project.toml:5]:
```suggestion
```
According to the [docs](https://julialang.github.io/Pkg.jl/v1/creating-packages/#Test-specific-dependencies-in-Julia-1.2-and-above-1): "Note that Pkg will add the tested package itself implictly."

**Comment by theogf** [comment]
All tests are passing, if there is no objections I am going to merge

**Comment by willtebbutt** [comment]
@theogf do you know why coverage has dropped?

**Comment by theogf** [comment]
@willtebbutt, yes, I needed to add some functionalities (mostly for Zygote) but sometimes it was not enough to make all tests pass (yet I left them there as they will be needed in the future). Right now tests are either all correct and removed. I want to tackle all of the different AD problems (all mentioned in #116 ) but would prefer to go one PR at a time, ss there are definitely a lot of issues.
